### ISSUE TITLE
CodeGen v2 codegen fails for zero length array

### ISSUE THREAD

**Issue by JakeHillion** [facebookexperimental/object-introspection#295]
JIT compiler error:
```
1925988647140263860.cc:138:7: error: base class has incomplete type
    : validate_size<Type, ExpectedSize, sizeof(Type)> {};
      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1925988647140263860.cc:213:15: note: in instantiation of template class 'validate_size<OIInternal::(anonymous namespace)::Foo0_0, 0, 0>' requested here
static_assert(validate_size<Foo0_0, 0>::value);
              ^
1925988647140263860.cc:132:8: note: definition of 'validate_size<OIInternal::(anonymous namespace)::Foo0_0, 0, 0>' is not complete until the closing '}'
struct validate_size : std::true_type {
       ^
1 error generated.
```

Unclear why this is happening. #294 does not fix it, and everything looks to be defined in the correct order. Needs further debugging.

Examples:
- `build/test/integration/integration_test_runner --gtest_filter='OidIntegration.arrays_member_int0' --verbose --preserve --enable-feature type-graph`

**Comment by ttreyer** [comment]
Why do we have both a 3-template-params version and a 2-template-param version that inherit from the first? Do we ever use the 3-template-params version directly?

I'm able to pass the static assert if I use two distinct structs (like `validate_size_3` and `validate_size_2`), instead of only one version that inherit from the other. Both Clang and GCC are unhappy about the inheritance, but fine with the two distinct structs.
See https://godbolt.org/z/9YWMvozof, I could update our implementation to use the two distinct structs, but I don't know if CodeGen will need updating.

**Comment by JakeHillion** [comment]
I see no problem with renaming the 3 param version, as you say it should only be used indirectly by the 2 param version. The reason there are two is to get the ActualSize explicitly in the 3 parameter version so it's included in the compiler error.

The best fix is probably to rename the two param version to `validate_size_eq` and remove the default value on `ActualSize`.

Thanks for looking at this!

**Comment by ttreyer** [comment]
Fixed by #453
